fw/ss_skb.c

```diff
@@ -1008,6 +1008,94 @@ ss_skb_chop_head_tail(struct sk_buff *skb_head, struct sk_buff *skb,
 	return 0;
 }
 
+/**
+ * Chop @head and @tail bytes at head and end of the message (skb circular
+ * list), iterating over the list if nessessary.
+ * @return 0 on success and -(error code) on error.
+ *
+ * Message length should be greater than the sum of @head and @trail,
+ * otherwise an error is retuned.
+ */
+int
+ss_skb_list_chop_head_tail(struct sk_buff **skb_list_head, 
+			   size_t head, size_t trail)
+{
+
+	struct sk_buff *skb, *skb_hd;
+	size_t sum;
+	int R;
+	skb = *skb_list_head;
+	if (likely(skb->next == skb))
+		goto single_buff;
+	if (likely(head))
+		goto head_and_probably_trail;
+	if (likely(trail))
+		goto trail_only;
+	/* nothing to chop */
+	return 0;
+
+single_buff:
+	/* There is the only 1 buffer in the list */
+	sum = head + trail;
+	if (WARN_ON(skb->len <= sum))
+		return -EBADMSG;
+	if (sum == 0)
+		/* Nothing to chop */
+		/* This check is mostly for jumps from branches below */
+		return 0;
+	return ss_skb_chop_head_tail(skb, skb, head, trail);
+
+head_and_probably_trail:
+	do {
+		if (likely(skb->len > head)) {
+			R = ss_skb_chop_head_tail(skb, skb, head, 0);
+			if (unlikely(R < 0))
+				return R;
+			if (unlikely(trail)) {
+				head = 0;
+				goto trail_only;
+			}
+			return 0;
+		}
+		head -= skb->len;
+		/* We do not use ss_skb_unlink() here to prevent it
+	 	* to remove the last skb in the list and to skip
+	 	* some actions using our apriori knowledge
+	 	*/
+		skb->next->prev = skb->prev;
+		skb->prev->next = skb->next;
+		*skb_list_head = skb_hd = skb->next;
+		__kfree_skb(skb);
+		skb = skb_hd;
+		if (unlikely(skb->next == skb))
+			goto single_buff;
+	} while (head);
+	if (!trail)
+		return 0;
```

Comment:
We don't need the check if we rewrite the loop for `trail` in precondition form

Reply:
@krizhanovsky 
But I still believe in the postcondition form.
In general - there are many redundant condition in the functions - textually,
but they are not redundant in run.

In particular, there is initialization action
```
skb_hd = *skb_list_head;
```
which should be done before loop and so independent of pre/post condition result,
currently I avoid it.

In even more general, we can produce more fast code or more maintenable code, the decision.
Here I tried to follow traditions shown in other funcs in ss_skb.c.

___
I see more comments on code comparison which do not appear here.
Moving to them...

Reply:
Fixed in new version